client/src/app/queries/sboms.ts

```diff
@@ -50,6 +53,24 @@ export const useFetchSBOMById = (id?: number | string) => {
   };
 };
 
+export const useDeleteSBOMByIdMutation = () => {
+  const queryClient = useQueryClient();
+  const {pushNotification} = React.useContext(NotificationsContext);
+
+  return useMutation({
+    mutationFn: (id: number | string) => deleteSBOMById(id),
+    onSuccess: (_res, id) => {
+      queryClient.invalidateQueries({queryKey: [SBOMsQueryKey, id]});
+    },
+    onError: (err: AxiosError, _payload: number | string) => {
+      pushNotification({
+        title: "Error while deleting SBOM",
+        variant: "danger",
+      });
```

Comment:
I agree and ideally we would like the notification to be passed as a callback.

Reply:
seems lime more boilerplate code for not much benefit.  What is the benefit for this hard separation?  hooks are meant to be composable so that you get a simple interface to complex behavior.  I see this a composed hook that is hiding the use of the notification context.  This really simplifies it use to the caller and provides the consistent behavior that is desired.  We should be aiming to reduce boilerplate code that get repeated.

Reply:
Oh and it's also not rendering DOM directly.  It's updating state in a context.

Reply:
I was working on a hook that would  on making these mutations more DRY by implementing a `useAppMutation` function that simplifies things down to:

```tsx
export const useDeleteProductByIdMutation = () => useAppMutation({
  mutationFn: (id: number | string) => deleteProductById(id),
  mutationKeyFn: (id: number | string) => [ProductsQueryKey, id],
  errorMessage: "Error while deleting Products",
});
```

Reply:
@chirino I've just create a PR to your branch here https://github.com/chirino/trustify-ui/pull/1

- The approach of setting DOM logic into the Queries is that is assumes there is only one way of deleting entities: "Delete the entity and then always show toastr notification". What usually happens as the app grows is something like:

 - Function1 calls deleteEntity - then show notification (what is PR is currently covering)
 - Function2 calls deleteEntity - do not show the notification but render the confirmation or error somewhere else like inside the modal itself

![error-and-validation](https://github.com/user-attachments/assets/70da6451-6f6d-4e09-be0a-d3f8cac3241c)

Decoupling data management (Queries) from UI (DOM) render actions helps to reuse data changes without impacting other parts of the UI.

Reply:
wouldn't we just add an option to the hook at that point to do the custom error handling?  Defaulting the current behaviour?

Reply:
Technically it would work. But the point is: not because we can do it means it is the best way of organizing things.
Look at the Queries Layer as you would look at the `services` layer in an backend application. I could easily mix business logic + tracking logic + pure REST logic, etc. under the `services` layer but although it works, it is not the way of organizing things in the long term. 

If you look at the PR I raised to your PR you will notice that it is the same code, the difference is that the Notifications is handled in the caller (the sbom-list page) and not in the Query (data management layer). Please change the code accordingly.

Reply:
So let me try to ask the question in a different way.   If I were to create a hook that simplified some of this UI behavior to reduce code duplication and reduce the chances that another part of the UI has inconsistent behavior, what directory would I store it in?

Reply:
- /client/src/app/hooks global scope hooks
- /client/src/app/components
- /client/src/app/queries data management

Reply:
which category would you put a hook like this one in?
 
```tsx
import type {DefaultError, QueryClient} from "@tanstack/query-core";
import {
  MutationKey,
  useMutation,
  UseMutationOptions,
  UseMutationResult,
  useQueryClient
} from "@tanstack/react-query";
import React from "react";
import {NotificationsContext} from "@app/components/NotificationsContext";

export function useAppMutation<
  TData = unknown,
  TError = DefaultError,
  TVariables = void,
  TContext = unknown,
>(
  options: UseMutationOptions<TData, TError, TVariables, TContext> & {
    errorMessage: string;
    mutationKeyFn?: (vars: TVariables, ctx: TContext) => MutationKey;
  },
  queryClient?: QueryClient
): UseMutationResult<TData, TError, TVariables, TContext> {

  const client = useQueryClient(queryClient);
  const {pushNotification} = React.useContext(NotificationsContext);
  return useMutation(
    {
      ...options,
      onSuccess: (res, vars, ctx) => {
        if (options.mutationKeyFn) {
          client.invalidateQueries({queryKey: options.mutationKeyFn(vars, ctx)});
        }
        if (options.onSuccess) {
          options.onSuccess(res, vars, ctx);
        }
      },
      onError: (err, vars, ctx) => {
        pushNotification({
          title: options.errorMessage,
          variant: "danger",
        });
        if (options.onError) {
          options.onError(err, vars, ctx);
        }
      },
    },
    queryClient
  );
}
```

Reply:
I don't see it matching any category. Personally, it looks like forcing things under the premise of "reduce code duplication".

Reply:
Or the categories should not be as rigid.  The point of components and hooks is to be able to do this kinda of code de-duplication in composable ways.  While I'm not a purist about removing all code duplication, it's generally a good thing as your code base gets bigger and you want an easy way to guarantee the same look and feel of different UI elements.